### ISSUE TITLE
Transition away from environment variables to a configuration file

### ISSUE THREAD

**Issue by ChipWolf** [itzg/docker-minecraft-server#1439]
### Enhancement Type

A completely new feature

### Describe the enhancement

The number of configuration options and environment variables is now quite large and unwieldy.
I think it's time to start consolidating everything down into a config file.

Additionally, the documentation for this container is likely worth moving to Mkdocs or similar to make it more navigable and accessible.

**Comment by itzg** [comment]
> The number of configuration options and environment variables is now quite large and unwieldy.
> 
> I think it's time to start consolidating everything down into a config file.


I appreciate the concern, but that is a huge undertaking and would be immensely disruptive to all of the existing users.

In any case compose files, deployment manifests, and helm values are already "config files". What advantages do you propose that a formal config file format would provide? 

> Additionally, the documentation for this container is likely worth moving to Mkdocs or similar to make it more navigable and accessible.

Please open a separate issue for this.

**Comment by ChipWolf** [comment]
@itzg does this issue need to be moved to the helper repo?

**Comment by itzg** [comment]
This is still the right place for it.

**Comment by ChipWolf** [comment]
> In any case compose files, deployment manifests, and helm values are already "config files". What advantages do you propose that a formal config file format would provide?

The primary benefit I see is in idempotency. For example; if it were viable to ship the proposed `.itzg-mc` config as part of generic pack builds, the variables could be iterated on and versioned declaratively.

Modpacks generally have a hard requirement on specific configuration - a file-based config could eliminate drift between the desired and actual state of the container from the pack's perspective.

For example, if a mod update requires a new version of Forge; the generic pack change to update the mod could also bump the Forge version.

**Comment by itzg** [comment]
The overall feature is starting to seem more reasonable. 

However

> Modpacks generally have a hard requirement on specific configuration - a file-based config could eliminate drift between the desired and actual state of the container from the pack's perspective.

This is already supported by some of the modpacks (not as generic packs) and I'm planning on broadening the modpack driven setup support.

So let's focus this particular issue/enhancement on non-server-type/version configuration.

**Comment by brandon1024** [comment]
@ChipWolf  Wouldn't the `--env-file` option of `docker-run` provide what you're looking for? Rather than configuring each environment variable through `-e <var>` options, you could maintain your own environment config file.

```
--env-file=file
       Read in a line delimited file of environment variables. See Environment note below for precedence.
```

It wouldn't hurt to keep a template file directly in the project with all supported variables and short descriptions for each variable.

**Comment by ChipWolf** [comment]
> @ChipWolf  Wouldn't the `--env-file` option of `docker-run` provide what you're looking for? Rather than configuring each environment variable through `-e <var>` options, you could maintain your own environment config file.
> 
> 
> 
> ```
> 
> --env-file=file
> 
>        Read in a line delimited file of environment variables. See Environment note below for precedence.
> 
> ```
> 
> 
> 
> It wouldn't hurt to keep a template file directly in the project with all supported variables and short descriptions for each variable.
> 
> 

Not entirely, I'm proposing a config file for the container, Docker is not the only container engine

**Comment by brandon1024** [comment]
What other OCI implementations are out there? :sweat_smile: I use podman but this flag is supported by every implementation I'm aware of (docker, podman and docker/podman-compose, and event k8s with config maps or secrets).

If you'd prefer not to use `--env-file`, another solution would be to have the entrypoint source an env-var file at startup.

```
-e SOURCE_ENV_FILE=true
-e ENV_FILE=/<some-mounted-file>
```

**Comment by itzg** [comment]
I think there's a lot of merit in @brandon1024 's suggestion. Certainly the command line syntax shown is specific to `docker run`, but every container platform I have worked with supports "environment variable files". That ubiquity isn't surprising given industry guidance like https://12factor.net/config .

Even without the container engine supporting it, the startup script could load an environment file like

```
set -a
source "${someEnvFile}"
set +a
```

I'm still not sure what a proprietary config file format would gain. A user starts with a blank file and then what? They need to refer to documentation like today. 

Are you envisioning a file that  the user grabs from a baseline copy that is pre populated with all possible variables and the user adjusts the ones that apply to their scenario?

Or is it yaml based in order to bring structure and organization to the configuration rather than a flat set of environment variables?

**Comment by itzg** [comment]
Oops, you beat me to the same point @brandon1024 😀 ...I kept getting distracted while writing out my comment.

**Comment by brandon1024** [comment]
@itzg haha no worries mate!

If it were up to me, I would steer clear of a yaml config file :sweat_smile:  To me, I see so much unnecessary complexity with very little gain in what @ChipWolf proposed. Yaml needs to be parsed meaning we would need to package a parser of some kind, and I wouldn't envy anyone that needs to write the (non-trivial) scripts to convert between _legacy_ variables and yaml.

And to to reformulate what @itzg hinted at, an env-var file is self-documenting; a yaml file would not be. I will agree that the configuration options of this container is a bit overwhelming at first glance, but I think that could be better solved with improved structured documentation. I digress :)

> Are you envisioning a file that the user grabs from a baseline copy that is pre populated with all possible variables and the user adjusts the ones that apply to their scenario?

I actually really like this idea; we could provide a baseline template env-var file at the project root similar to a handful of projects out there ([dnsmasq](https://github.com/imp/dnsmasq/blob/master/dnsmasq.conf.example), [kafka](https://github.com/apache/kafka/blob/trunk/config/server.properties), [redis](https://github.com/redis/redis/blob/unstable/redis.conf), [sshd](https://github.com/openssh/openssh-portable/blob/master/sshd_config), etc).

**Comment by ChipWolf** [comment]
The major benefit I can see from using something like a yaml config is due to the feature hierarchy.

Variables are relatively linear, but given the number of variables defined for each feature the structure of a a hierarchical config would be nice.

```
FOO: true
FOO_BAR: true
FOO_BAR_BAZ: banana
```

```yaml
foo:
  enabled: true
  bar:
    enabled: true
    baz: banana
```

As long as there's consistent naming scheme for variables, a transformer between the two would be universal.

---

Imagine you have a pack distributed for use with the itzg image, you'd like to ship it with some defaults, the pack itself could contain those defaults.
The user could choose to override them with variables if they wished.

Right now, a change to the modpack (one artifact) might be a breaking change to your infrastructure (another artifact).

Let's imagine you added a config variable to be used with `REPLACE_ENV`, the addition of a single variable to the pack is now a major version bump because the consumer of the pack must make a change to their own configuration to support the change.

If a default config was included in the pack, this addition could easily be a patch/minor version bump which wouldn't require a consumer of the pack to change anything.

**Comment by brandon1024** [comment]
@ChipWolf if you really want to configure the server with yaml-based configuration, maybe consider creating a project that uses this project as a base layer and provides the functionality you're looking for :-)

**Comment by ChipWolf** [comment]
> @ChipWolf if you really want to configure the server with yaml-based configuration, maybe consider creating a project that uses this project as a base layer and provides the functionality you're looking for :-)

Happy continuing contributing here, itzg has poured enormously admirable effort into a phenomenal resource & it's very well maintained; a fork would be crazy overhead.

I'd be happy to PR anything the community lands on but I wasn't wanting to spend hours on a huge refactor before an RFC.

I'm happy with an env-file approach if we're not wanting to lift out more of the logic into the wrapper/helper anytime soon

**Comment by brandon1024** [comment]
> a fork would be crazy overhead.

Apologies, maybe my wording wasn't super great. I didn't mean to suggest a fork of the project (agreed, forking would be a nightmare), but rather just a docker image that is based on `itzg/minecraft-server` that includes your yaml configuration capabilities:

```
FROM itzg/minecraft-server

COPY <your configuration scripts>

ENTRYPOINT [ "<your entrypoint>" ]
```

**Comment by itzg** [comment]
> I imagine as more of the complexity was being dissolved out of the shell scripts into the mc-image-helper, the vision is to continue in the direction of [mostly] eliminating the scripts at some point?

Yes. In fact, I could backpedal a little bit and point out that support yaml driven configs in mc-image-helper is actually quite doable, but still an additional maintenance vector I wouldn't want to maintain myself.

FWIW, @brandon1024 I knew what you meant about leveraging the base image 😀 and agree that is the kind of "frontend" that I could see enabling a conversion to environment variables.

**Comment by ChipWolf** [comment]
> an additional maintenance vector

I'd be happy to maintain something like this, but with the proposed design here, it would become the only vector for maintaining variables - there wouldn't be any additional overhead so to speak.

There'd of course be a one-time effort to plug a config handler into the helper app and map the "legacy" variables to the new format; but after that point, any new variables would be exposed automatically per the LuckPerms example above and could still be consumed by the existing scripts.

The scripts would reference the environment variables directly as they do now, with a view of the helper apps gradually absorbing the scripts.

**Comment by itzg** [comment]
I appreciate the offer, but I really don't want any part of that in any of the repos I maintain. These are my projects and I have particular design strategies throughout -- this basically changes everything. I'm sorry if I misled you by entertaining the initial discussion on this topic. 

You're welcome to implement all that in adjacent tooling, forked repos, extending the image, etc. 

I'm still living with the mistake of owning the helm charts (reminded by the latest issue report there) and I am already spending too much of my personal time on all of these things with very little compensation.

**Comment by itzg** [comment]
I will offer this concession: 

I have been needing to maintain (and will build a tool to generate) a manifest of all environment variables used by the image, their defaults, category, and brief description when feasible. I'll either produce a json or yaml file that will be committed and baked into the image. I might add an `mc-image-helper` command to dump the file in various formats: json, yaml, CSV, markdown table, etc.
